test/EndToEnd/Ledger.hs

```diff
@@ -208,3 +226,250 @@ unauthorisedBidderTest = mkAssertion $ do
+
+  nftTx <- mintOneTestNFT
+  let utxoRef = mkTxIn nftTx 0
+
+  terms <- liftIO $ do
+    dynamicState <- constructTermsDynamic seller utxoRef
+    configToAuctionTerms config dynamicState
+
+  assertNFTNumEquals seller 1
+
+  announceAuction terms
+
+  withActor buyer1 $ mkDeposit terms
+  withActor buyer2 $ mkDeposit terms
+
+  assertUTxOsInScriptEquals Deposit terms 2
+
+  waitUntil $ biddingStart terms
+  actorsPkh <- liftIO $ getActorsPubKeyHash [buyer1, buyer2]
+  startBidding terms (ApprovedBidders actorsPkh)
+
+  assertNFTNumEquals seller 0
+
+  withActor buyer1 $ newBid terms $ startingBid terms
+  withActor buyer2 $ newBid terms $ startingBid terms + minimumBidIncrement terms
+
+  waitUntil $ biddingEnd terms
+
+  withActor buyer1 $ losingBidderClaimDeposit terms
+
+  assertUTxOsInScriptEquals Deposit terms 1
+
+  result <- try $ withActor buyer1 $ losingBidderClaimDeposit terms
+
+  case result of
+    Left (_ :: SomeException) -> assertUTxOsInScriptEquals Deposit terms 1
+    Right _ -> fail "User should not be able to claim depoist twice"
+
+sellerClaimsDepositTest :: Assertion
+sellerClaimsDepositTest = mkAssertion $ do
+  let seller = Alice
+      buyer = Bob
+
+  mapM_ (initWallet 100_000_000) [seller, buyer]
+
+  nftTx <- mintOneTestNFT
+  let utxoRef = mkTxIn nftTx 0
+
+  terms <- liftIO $ do
+    dynamicState <- constructTermsDynamic seller utxoRef
+    configToAuctionTerms config dynamicState
+
+  assertNFTNumEquals seller 1
+
+  announceAuction terms
+
+  withActor buyer $ mkDeposit terms
+
+  assertUTxOsInScriptEquals Deposit terms 1
+
+  waitUntil $ biddingStart terms
+  actorsPkh <- liftIO $ getActorsPubKeyHash [buyer]
+  startBidding terms (ApprovedBidders actorsPkh)
+
+  assertNFTNumEquals seller 0
+
+  withActor buyer $ newBid terms $ startingBid terms
+
+  waitUntil $ voucherExpiry terms
+
+  sellerClaimDepositFor terms (head actorsPkh)
+
+  assertUTxOsInScriptEquals Deposit terms 0
+
+sellerClaimsLosingDepositTest :: Assertion
+sellerClaimsLosingDepositTest = mkAssertion $ do
+  let seller = Alice
+      buyer1 = Bob
+      buyer2 = Carol
+
+  mapM_ (initWallet 100_000_000) [seller, buyer1, buyer2]
+
+  nftTx <- mintOneTestNFT
+  let utxoRef = mkTxIn nftTx 0
+
+  terms <- liftIO $ do
+    dynamicState <- constructTermsDynamic seller utxoRef
+    configToAuctionTerms config dynamicState
+
+  assertNFTNumEquals seller 1
+
+  announceAuction terms
+
+  withActor buyer1 $ mkDeposit terms
+  withActor buyer2 $ mkDeposit terms
+
+  assertUTxOsInScriptEquals Deposit terms 2
+
+  waitUntil $ biddingStart terms
+  actorsPkh <- liftIO $ getActorsPubKeyHash [buyer1, buyer2]
+  startBidding terms (ApprovedBidders actorsPkh)
+
+  assertNFTNumEquals seller 0
+
+  withActor buyer1 $ newBid terms $ startingBid terms
+  withActor buyer2 $ newBid terms $ startingBid terms + minimumBidIncrement terms
+
+  waitUntil $ voucherExpiry terms
+
+  result <- try $ sellerClaimDepositFor terms (head actorsPkh)
+
+  case result of
```

Comment:
I think that could be merged with `sellerClaimsDepositTest` cuz we have both winning and losing bidder here.

Reply:
Okay.

My thinking was that checking everything separately in E2E tests may lead to case explosion.